tools/cldr/cldr-to-icu/README.md

```diff
@@ -0,0 +1,140 @@
+pre-processed locally into a different directory.
+
+
+## Initial Setup
+
+This project relies on the Maven build tool for managing dependencies and uses
+Ant for configuration purposes, so both will need to be installed. On a Debian
+based system, this should be as simple as:
+
+```
+$ sudo apt-get install maven ant
+```
+
+You must also install an additional CLDR JAR file the local Maven repository at
+`$TOOLS_ROOT/cldr/lib` (see the README.txt in that directory for more
+information).
+
+```
+$ cd "$TOOLS_ROOT/cldr/lib"
+$ ./install-cldr-jars.sh "$CLDR_DIR"
+```
+
+## Generating all ICU data and source code
+
+```
+$ cd "$TOOLS_ROOT/cldr/cldr-to-icu"
+$ ant -f build-icu-data.xml
+```
+
+## Other Examples
+
+* Outputting a subset of the supplemental data into a specified directory:
+  ```
+  $ ant -f build-icu-data.xml -DoutDir=/tmp/cldr -DoutputTypes=plurals,dayPeriods -DdontGenCode=true
+  ```
+  Note: Output types can be listed with mixedCase, lower_underscore or UPPER_UNDERSCORE.
+  Pass `-DoutputTypes=help` to see the full list.
+
+
+* Outputting only a subset of locale IDs (and all the supplemental data):
+  ```
+  $ ant -f build-icu-data.xml -DoutDir=/tmp/cldr -DlocaleIdFilter='(zh|yue).*' -DdontGenCode=true
+  ```
+
+* Overriding the default CLDR version string (which normally matches the CLDR library code):
+  ```
+  $ ant -f build-icu-data.xml -DcldrVersion="36.1"
+  ```
+
+* Using alternate CLDR values (ex: use `alt="ascii"` values from the CLDR XML):
+
+  First, edit the `build-icu-data.xml` file where it mentions "ALTERNATE VALUES"
+  with the correctly annotated source path, target path, and locales list:
+  ```diff
+  @@ -384,6 +399,20 @@
+            <!-- ALTERNATE VALUES -->
+
+            <!-- The following elements configure alternate values for some special case paths.
+                 The target path will only be replaced if both it, and the source path, exist in
+                 the CLDR data (paths will not be modified if only the source path exists).
+
+                 Since the paths must represent the same semantic type of data, they must be in the
+                 same "namespace" (same element names) and must not contain value attributes. Thus
+                 they can only differ by distinguishing attributes (either added or modified).
+
+                 This feature is typically used to select alternate translations (e.g. short forms)
+                 for certain paths. -->
+             <!-- <altPath target="//path/to/value[@attr='foo']"
+                           source="//path/to/value[@attr='bar']"
+                           locales="xx,yy_ZZ"/> -->
+  +            <altPath target="//ldml/dates/calendars/calendar[@type='all']/dateTimeFormats/availableFormats/dateFormatItem[@alt='ascii']"
+  +                     source="//ldml/dates/calendars/calendar[@type='all']/dateTimeFormats/availableFormats/dateFormatItem"
+  +                     locales="en"/>
+  ```
```

Comment:
Done.

Btw, is the `[@type='all']` syntax correct? I thought I heard you say it that way, so I wrote it. But I didn't know where to go to verify valid syntax for the tool.

Reply:
@pedberg-icu @macchiati could you please check the CLDR XPaths here?

Reply:
@pedberg-icu @macchiati any chance for a look at the CLDR XPaths here before Elango merges?

Reply:
@hagbard is it possible to use wildcards in CLDR XPaths so that we can select alternate data values (in this case, for `alt="ascii"` on several dozen items) without enumerating every possible path? Please see the changes and comments in this pull request.

@pedberg-icu @macchiati if you know the answer or have a list of paths... please chime in!

Reply:
@markusicu @echeran @mihnita  I do not normally use these instructions, I use the ones in icu4c/source/data/cldr-icu-readme.txt. However as far as the paths are concerned, please look at the normal ones in tools/cldr/cldr-to-icu/src/main/resources/ldml2icu_locale.txt; there is a wildcard mechanism there. I do not know whether the altPaths can use the variables defined there such as %A.

As far as the paths we care about - I think we can focus only on gregorian calendar, so (using the %A wildcard notation here):
```
//ldml/dates/calendars/calendar[@type="gregorian"]/timeFormats/timeFormatLength[@type="(%A)"]/timeFormat[@type="%A"][@alt="ascii"]/pattern[@type="%A"]
//ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats/dateFormatItem[@id="(%A)"][@alt="ascii"]
```

Reply:
Thank you.

I've updated the `tools/cldr/cldr-to-icu/build-icu-data.xml` file to use the `%A`  wildcard as above.
And unfortunately the generated text files still contain `U+202F` instead of the expected U+0020.
So it looks like the wildcard does not work in `altPath` :-(

This is what the rules I added looked like:

```xml
<altPath target="//ldml/dates/calendars/calendar[@type='gregorian']/dateTimeFormats/availableFormats/dateFormatItem[@id='(%A)']"
         source="//ldml/dates/calendars/calendar[@type='gregorian']/dateTimeFormats/availableFormats/dateFormatItem[@id='(%A)'][@alt='ascii']"
         locales="en"/>                                                                             
<altPath target="//ldml/dates/calendars/calendar[@type='gregorian']/timeFormats/timeFormatLength[@type='(%A)']/timeFormat[@type='%A']/pattern[@type='%A']"
         source="//ldml/dates/calendars/calendar[@type='gregorian']/timeFormats/timeFormatLength[@type='(%A)']/timeFormat[@type='%A'][@alt='ascii']/pattern[@type='%A']"
         locales="en"/>                                                                             
``` 

The only difference is that I replaced `"` with `'` (because the xpath is already in a `"`-delimited string),
and removed `[@alt='ascii']` from the `target`

Reply:
Do you think it would be easier to add this kind of customization in `tools/cldr/cldr-to-icu/src/main/resources/ldml2icu_locale.txt` ?

Thank you,
M

Reply:
Note sure how adding it in ldml2icu_locale.txt would work. But instead of adding %A for a wildcard, why not use what %A is defined to be in ldml2icu_locale.txt, i.e. [^"']++

Reply:
Peter, we might want to also include the generic calendar. I agree that we shouldn't use the others. That is:

Definitely:
```
//ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats/dateFormatItem[@id="Ehm"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats/dateFormatItem[@id="Ehms"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats/dateFormatItem[@id="h"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats/dateFormatItem[@id="hm"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats/dateFormatItem[@id="hms"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats/dateFormatItem[@id="hmsv"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats/dateFormatItem[@id="hmv"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="gregorian"]/timeFormats/timeFormatLength[@type="full"]/timeFormat[@type="standard"]/pattern[@alt="ascii"][@type="standard"]
//ldml/dates/calendars/calendar[@type="gregorian"]/timeFormats/timeFormatLength[@type="long"]/timeFormat[@type="standard"]/pattern[@alt="ascii"][@type="standard"]
//ldml/dates/calendars/calendar[@type="gregorian"]/timeFormats/timeFormatLength[@type="medium"]/timeFormat[@type="standard"]/pattern[@alt="ascii"][@type="standard"]
//ldml/dates/calendars/calendar[@type="gregorian"]/timeFormats/timeFormatLength[@type="short"]/timeFormat[@type="standard"]/pattern[@alt="ascii"][@type="standard"]
```

Perhaps:

```
//ldml/dates/calendars/calendar[@type="generic"]/dateTimeFormats/availableFormats/dateFormatItem[@id="Ehm"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="generic"]/dateTimeFormats/availableFormats/dateFormatItem[@id="Ehms"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="generic"]/dateTimeFormats/availableFormats/dateFormatItem[@id="h"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="generic"]/dateTimeFormats/availableFormats/dateFormatItem[@id="hm"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="generic"]/dateTimeFormats/availableFormats/dateFormatItem[@id="hms"][@alt="ascii"]

```

NOT:
```
//ldml/dates/calendars/calendar[@type="chinese"]/dateTimeFormats/availableFormats/dateFormatItem[@id="Ehm"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="chinese"]/dateTimeFormats/availableFormats/dateFormatItem[@id="Ehms"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="chinese"]/dateTimeFormats/availableFormats/dateFormatItem[@id="h"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="chinese"]/dateTimeFormats/availableFormats/dateFormatItem[@id="hm"][@alt="ascii"]
//ldml/dates/calendars/calendar[@type="chinese"]/dateTimeFormats/availableFormats/dateFormatItem[@id="hms"][@alt="ascii"]
//ldml/localeDisplayNames/types/type[@key="d0"][@type="ascii"]
//ldml/localeDisplayNames/types/type[@key="s0"][@type="ascii"]
```

Reply:
That set (including the perhaps) seems good, and avoids the wildcards.

Reply:
+1 to explicit rules (until we manage to fix the code)

I tried the wildcard (`%A`), and tried the explicit definition (` [^"']++`), and a simpler one (` [A-Za-z]++`).
And nothing works.

I don't think the problem is the xpath selection itself.

These xpaths (without attributes, or with attributes with wildcard values) all work and select many nodes.
And the conversion tool seems to not be able to handle many nodes, only 1 returned from `source` and 1 returned from `target`. 
The selector with `ascii` returns ~17 nodes, the one without about 180.
So the tool seem confused: "how can I map 17 nodes to 180"?

That is why the explicit rules work, because they result in 1:1 mapping.
They don't scale (need to be updated every time we add a new entry with `alt='ascii'`), but it is the best we can do without a code update.

Reply:
Is the second plus sign in `[^"']++` intentional?

@echeran is adding the explicit paths in PR #2503. @mihnita could you please try to run the converter with those?